WebDriverAgentLib/Utilities/FBXPath.m

```diff
@@ -0,0 +1,270 @@
+/**
+ * Copyright (c) 2015-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+#import "FBXPath.h"
+
+#import "FBLogger.h"
+#import "XCAXClient_iOS.h"
+#import "XCTestDriver.h"
+#import "XCTestPrivateSymbols.h"
+#import "XCUIElement.h"
+#import "XCUIElement+FBWebDriverAttributes.h"
+
+const static char *_UTF8Encoding = "UTF-8";
+
+static NSString *const kXMLIndexPathKey = @"private_indexPath";
+static NSString *const topNodeIndexPath = @"top";
+NSString *const XCElementSnapshotInvalidXPathException = @"XCElementSnapshotInvalidXPathException";
+NSString *const XCElementSnapshotXPathQueryEvaluationException = @"XCElementSnapshotXPathQueryEvaluationException";
+
+@implementation FBXPath
+
++(void)throwException:(NSString *)name forQuery:(NSString *)xpathQuery __attribute__((noreturn))
+{
+  NSString * reason = [NSString stringWithFormat:@"Cannot evaluate results for XPath expression \"%@\"", xpathQuery];
+  @throw [NSException exceptionWithName:name reason:reason userInfo:@{}];
+}
+
++ (NSArray<XCElementSnapshot *> *)findMatchesIn:(XCElementSnapshot *)root withXPathQuery:(NSString *)xpathQuery
+{
+  int rc;
+  xmlTextWriterPtr writer;
+  xmlDocPtr doc;
+  
+  writer = xmlNewTextWriterDoc(&doc, 0);
+  if (NULL == writer) {
+    [FBLogger logFmt:@"Failed to invoke libxml2>xmlNewTextWriterDoc for XPath query \"%@\"", xpathQuery];
+    [FBXPath throwException:XCElementSnapshotXPathQueryEvaluationException forQuery:xpathQuery];
```

Comment:
Can you add return after each exception throw? I know it doesn't change any thing but makes is easier to read and "scan".